src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -166,139 +244,109 @@ public static Response Apply(Response current, Stack<Request> requests, List<Res
     }
 
     internal class ReplaceTypeParamCallsExpression :
-        ExpressionTransformation<ReplaceTypeParamCallsExpressionKind>
+        ExpressionTransformation<ReplaceTypeParamCallsExpressionKind, ReplaceTypeParamCallsExpressionType>
     {
-        private Response CurrentResponse;
-        private Stack<Request> Requests;
-        private List<Response> Responses;
+        private readonly Dictionary<QsTypeParameter, ResolvedType> CurrentParamTypes;
 
-        public ReplaceTypeParamCallsExpression(Response currentResponse, Stack<Request> requests, List<Response> responses) :
-            base(ex => new ReplaceTypeParamCallsExpressionKind(ex as ReplaceTypeParamCallsExpression))
+        public ReplaceTypeParamCallsExpression(Dictionary<QsTypeParameter, ResolvedType> currentParamTypes, GetConcreteIdentifierFunc getConcreteIdentifier) :
+            base(ex => new ReplaceTypeParamCallsExpressionKind(ex as ReplaceTypeParamCallsExpression, currentParamTypes, getConcreteIdentifier),
+                ex => new ReplaceTypeParamCallsExpressionType(ex as ReplaceTypeParamCallsExpression, currentParamTypes))
         {
-            CurrentResponse = currentResponse;
-            Requests = requests;
-            Responses = responses;
+            CurrentParamTypes = currentParamTypes;
         }
 
         public override TypedExpression Transform(TypedExpression ex)
         {
-            // TODO: This should be recursive and use a more robust data structure to handle the type information for generics
-            ImmutableDictionary<QsTypeParameter, ResolvedType> types = ex.TypeParameterResolutions;
+            var range                = this.onRangeInformation(ex.Range);
+            var typeParamResolutions = this.onTypeParamResolutions(ex.TypeParameterResolutions);
+            var exType               = this.Type.Transform(ex.ResolvedType);
+            var inferredInfo         = this.onExpressionInformation(ex.InferredInformation);
+            // Change the order so that Kind is transformed last
+            var kind                 = this.Kind.Transform(ex.Expression);
```

Comment:
Might be good to add a comment here saying that the order of recursion matters (onTypeParameterResolutions before the kind transformation). And it would definitively be good to add that on the todo list for test: write a test that fails if the order of these two is switched.